<h1>UK ROAD ACCIDENTS DATA ANALYTICS PROJECT</h1>

<h2>Analyst: Jovan Allyn L. Diaz</h2>

In [2]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [3]:
ukaccidents = pd.read_csv('datasets\\ukaccidents.csv')

In [4]:
ukaccidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>DataFrame Info</h1>

In [5]:
ukaccidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h1>DataFrame Basic Descriptive Statistics</h1>

In [6]:
ukaccidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

<h1>Checking Null Values</h1>

In [7]:
ukaccidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1>Fill Null Values with Statistical Values</h1>

In [8]:
ukaccidents['Latitude'] = ukaccidents['Latitude'].fillna(ukaccidents['Latitude'].mean())
ukaccidents['Longitude'] = ukaccidents['Longitude'].fillna(ukaccidents['Longitude'].mean())
ukaccidents['Road_Surface_Conditions'] = ukaccidents['Road_Surface_Conditions'].fillna(ukaccidents['Road_Surface_Conditions'].mode()[0])
ukaccidents['Road_Type'] = ukaccidents['Road_Type'].fillna(ukaccidents['Road_Type'].mode()[0])
ukaccidents['Urban_or_Rural_Area'] = ukaccidents['Urban_or_Rural_Area'].fillna(ukaccidents['Urban_or_Rural_Area'].mode()[0])
ukaccidents['Weather_Conditions'] = ukaccidents['Weather_Conditions'].fillna(ukaccidents['Weather_Conditions'].mode()[0])

<h1>Checking Column Data Types</h1>

In [9]:
ukaccidents.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [10]:
ukaccidents


Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Activity 2</h1>

<h1>Question 1</h1>

<h2>What is the most common accident severity?</h2>

In [11]:
ukaccidents['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h2>Insights: Most accidents are slight, showing that many are minor.</h2>

<h1>Question 2</h1>

<h2>What is the average number of casualties?</h2>

In [12]:
ukaccidents['Number_of_Casualties'].mean()

np.float64(1.357040257068864)

<h2>Insights: On Average, about 1 person is injured per accident.</h2>

<h1>Question 3</h1>

<h2>What is the maximun casualties in a single accidents?</h2>

In [13]:
ukaccidents['Number_of_Casualties'].max()

np.int64(68)

<h2>Insights: The worst accidents involve many casualties, but it is rare.</h2>

<h1>Question 4</h1>

<h2>What is the average number of vehicles involved?</h2>

In [14]:
ukaccidents['Number_of_Vehicles'].mean()

np.float64(1.8312554205597575)

<h2>Insights: Most accidents involve 1-2 vehicles.
</h2>

<h1>Question 5</h1>

<h2>What is the most common road surface conditon?</h2>

In [15]:
ukaccidents['Road_Surface_Conditions'].value_counts().head(1)

Road_Surface_Conditions
Dry    448547
Name: count, dtype: int64

<h2>Insights: Most happen on dry roads, since they are the most common condition.
</h2>

<h1>Question 6</h1>

<h2>What is the most common weather condition?</h2>

In [16]:
ukaccidents['Weather_Conditions'].value_counts().head(1)

Weather_Conditions
Fine no high winds    535013
Name: count, dtype: int64

<h2>Insights: Fine weather has the most accidents, since people drive more often.</h2>

<h1>Question 7</h1>

<h2>Where do more accidents happens, Urban or Rural areas?</h2>

In [17]:
ukaccidents['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>Insights: Urban areas have more accidents due to higher traffic.</h2>

<h1>Question 8</h1>

<h2>What is the most common light condition?</h2>

In [18]:
ukaccidents['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2>Insights: More accidents happen in daylight, but night accidents tend to be worse</h2>

<h1>Question 9</h1>

<h2>What is the most common vehicle type?</h2>

In [19]:
ukaccidents['Vehicle_Type'].value_counts().head(5)

Vehicle_Type
Car                                     497992
Van / Goods 3.5 tonnes mgw or under      34160
Bus or coach (17 or more pass seats)     25878
Motorcycle over 500cc                    25657
Goods 7.5 tonnes mgw and over            17307
Name: count, dtype: int64

<h2>Insights: Cars are the most common vehicles in accidents.</h2>

<h1>Question 10</h1>

<h2>What is the average casualties by vehicle type?
</h2>

In [20]:
ukaccidents.groupby('Vehicle_Type')['Number_of_Casualties'].mean().head(5)

Vehicle_Type
Agricultural vehicle                    1.342065
Bus or coach (17 or more pass seats)    1.349216
Car                                     1.358841
Data missing or out of range            1.500000
Goods 7.5 tonnes mgw and over           1.351881
Name: Number_of_Casualties, dtype: float64

<h2>Insights: Larger vehicles like buses may have higher average casualties.</h2>

<h1>Question 11</h1>

<h2>What is the most common road type?</h2>

In [21]:
ukaccidents['Road_Type'].value_counts().head(5)

Road_Type
Single carriageway    496663
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<h2>Insights: Single carriageways are the most accident-prone.</h2>

<h1>Question 12</h1>

<h2>What is the District with most accidents?</h2>

In [22]:
ukaccidents['District Area'].value_counts().head(1)

District Area
Birmingham    13491
Name: count, dtype: int64

<h2>Insights: Some districs have much higher accident counts, showing hostspots.</h2>

<h1>Question 13</h1>

<h2>What is the number of accidents with >2 vehicles?</h2>

In [23]:
(ukaccidents['Number_of_Vehicles']>2).sum()

np.int64(67897)

<h2>Insights: Multi-vehicle crashes are less common but dangerous.</h2>

<h1>Question 14</h1>

<h2>What is the number of accidents with >3 casualties?</h2>

In [24]:
(ukaccidents['Number_of_Casualties']>3).sum()

np.int64(17891)

<h2>Insights: Most accidents have few casualties, but some major events exist.</h2>

<h1>Question 15</h1>

<h2>What is the count of accidents on dry vs wet roads?</h2>

In [25]:
ukaccidents['Road_Surface_Conditions'].value_counts().head(2)

Road_Surface_Conditions
Dry            448547
Wet or damp    186708
Name: count, dtype: int64

<h2>Insights: Dry roads see the most accidents, but wet roads may be more dangerous.</h2>

<h1>Question 16 </h1>

<h2>What is the count of accidents in fine vs rainy weather?</h2>

In [26]:
ukaccidents['Weather_Conditions'].value_counts().head(2)

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Name: count, dtype: int64

<h2>Insights: Fine weather dominates, but rain still causes many accidents.</h2>

<h1>Question 17</h1>

<h2>What are the Districs with more than 10.000 accidents?
</h2>

In [27]:
ukaccidents['District Area'].value_counts()[ukaccidents['District Area'].value_counts()>10000]

District Area
Birmingham    13491
Name: count, dtype: int64

<h2>Insights: Birmingham have extremely high accident counts.</h2>

<h1>Question 18</h1>

<h2>What is the average casualties in Urban vs Rural areas?</h2>

In [28]:
ukaccidents.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h2>Insights: Rural accidents often have higher casualties, due to their higher speeds.</h2>

<h1>Question 19</h1>

<h2>What is the year with the most accidents?</h2>

In [29]:
ukaccidents['Accident Date'].value_counts().head(5)

Accident Date
30-11-2019    704
31-01-2019    697
13-07-2019    692
13-11-2021    692
14-08-2019    688
Name: count, dtype: int64

<h2>Insights: Accident numbers can show trends over time, it may increase or decrease over time.</h2>

<h1>Question 20</h1>

<h2>What is the average vehicles by severity?</h2>

In [30]:
ukaccidents.groupby('Accident_Severity')['Number_of_Vehicles'].mean()

Accident_Severity
Fatal      1.786976
Serious    1.678327
Slight     1.855864
Name: Number_of_Vehicles, dtype: float64

<h2>Insights: Serious or Fatal accidents may involve more vehicles on average, while Slight accidents are often single-vehicle or two-vehicle crashes</h2>

<h1>Clearing any Inconsistencies with the Data Set</h1>

In [32]:
ukaccidents['Accident Date'] = ukaccidents['Accident Date'].str.strip() #to remove any extra white spaces
ukaccidents['Accident Date'] = ukaccidents['Accident Date'].astype('str') #to convert into a string
ukaccidents['Accident Date'] = ukaccidents['Accident Date'].str.replace('/','-')

In [34]:
ukaccidents['Accident Date'] = pd.to_datetime(ukaccidents['Accident Date'], dayfirst=True, errors ='coerce')

<h1>Extracting Date Information using pandas date time</h1>

In [38]:
ukaccidents['Year'] = ukaccidents['Accident Date'].dt.year
ukaccidents['Month'] = ukaccidents['Accident Date'].dt.month
ukaccidents['Day'] = ukaccidents['Accident Date'].dt.day
ukaccidents['DayOfWeek'] = ukaccidents['Accident Date'].dt.dayofweek #monday=0, Sunday=6

In [39]:
ukaccidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

<h1>Question</h1>

<h2>Which road types have the most vehicles involved in accidents?</h2>

In [40]:
ukaccidents.groupby('Road_Type')['Number_of_Vehicles'].mean().sort_values(ascending=False).head(5)

Road_Type
Dual carriageway      2.009766
Slip road             1.917483
Roundabout            1.901323
Single carriageway    1.794803
One way street        1.585441
Name: Number_of_Vehicles, dtype: float64

<h2>Insights:
<li> Dual carriageways average ~2 vehicles per accident (highest).</li>
<li>Slip roads and roundabouts also have high averages.</li>
<li>Single carriageways and one-way streets involve fewer vehicles.</li>
<li>Busier/multi-lane roads lead to more multi-vehicle accidents.</li>


<h2>On average, how many casualties happen in each type of accident?</h2>

In [41]:
ukaccidents.groupby('Accident_Severity')['Number_of_Casualties'].mean()

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h2>Insights:
<li>Fatal accidents involve almost 2 people on average.</li>
<li>Serious accidents: about 1.47 people.</li>
<li>Slight accidents: about 1.33 people.</li>
<li>Fatal accidents usually affect more people per incident.</li>
</h2>